data/src/main/java/com/linkedin/data/codec/symbol/DefaultSymbolTableProvider.java

```diff
@@ -0,0 +1,137 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.data.codec.symbol;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.linkedin.data.codec.ProtobufCodecOptions;
+import com.linkedin.data.codec.ProtobufDataCodec;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * A default {@link SymbolTableProvider} implementation that doesn't use symbol tables for requests/responses of its
+ * own, but is able to retrieve remote symbol tables to decode responses from other services
+ */
+public class DefaultSymbolTableProvider implements SymbolTableProvider
+{
+  /**
+   * Metadata extractor
+   */
+  private static final SymbolTableMetadataExtractor METADATA_EXTRACTOR = new SymbolTableMetadataExtractor();
+
+  /**
+   * Accept header
+   */
+  private static final String ACCEPT_HEADER = "Accept";
+
+  /**
+   * Logger.
+   */
+  private static final Logger LOGGER = LoggerFactory.getLogger(DefaultSymbolTableProvider.class.getSimpleName());
+
+  /**
+   * Codec
+   */
+  static final ProtobufDataCodec CODEC =
+      new ProtobufDataCodec(new ProtobufCodecOptions.Builder().setEnableASCIIOnlyStrings(true).build());
+
+  /**
+   * Path from which symbol tables are served by remote Rest.li services.
+   */
+  public static final String SYMBOL_TABLE_URI_PATH = "symbolTable";
+
+  /**
+   * Cache storing mapping from symbol table name to symbol table.
+   */
+  final Cache<String, SymbolTable> _cache;
+
+  /**
+   * Constructor
+   */
+  DefaultSymbolTableProvider()
+  {
+    _cache = Caffeine.newBuilder().maximumSize(1000).build();
+  }
+
+  @Override
+  public SymbolTable getSymbolTable(String symbolTableName)
+  {
+    try
+    {
+      SymbolTableMetadata metadata = METADATA_EXTRACTOR.extractMetadata(symbolTableName);
+      String serverNodeUri = metadata.getServerNodeUri();
+      String tableName = metadata.getSymbolTableName();
+      boolean isRemote = metadata.isRemote();
+
+      // First check the cache.
+      SymbolTable symbolTable = _cache.getIfPresent(tableName);
+      if (symbolTable != null)
+      {
+        return symbolTable;
+      }
+
+      // If this is not a remote table, and we didn't find it in the cache, cry foul.
+      if (!isRemote)
+      {
+        throw new IllegalStateException("Unable to fetch symbol table with name: " + symbolTableName);
+      }
+
+      // Ok, we didn't find it in the cache, let's go query the service the table was served from.
+      String url = serverNodeUri + "/" + SYMBOL_TABLE_URI_PATH + "/" + tableName;
+      HttpURLConnection connection = openConnection(url);
```

Comment:
minor: disconnect finally since this is only one use.

Reply:
Good catch. Without this we would have leaked sockets on error responses. Fixed.